### PR TITLE
Calling MapLoader.loadAllKeys reflectively to preserve backwards compatibility

### PR DESCRIPTION
**Pull request by ajermakovics**
Issue #4371

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java [hunk 1]
```diff
@@ -99,7 +100,8 @@ public void deleteAll(Collection keys) {
 
     public Iterable loadAllKeys() {
         if (isMapLoader()) {
-            return mapLoader.loadAllKeys();
+            // Invoke reflectively to preserve backwards binary compatibility
+            return ReflectionHelper.invokeMethod(mapLoader, "loadAllKeys");
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java]: it should be just a fallback. 
calling a method reflectively (setAccessible(), etc) can cause funky issues when eg. SecurityManager is present.

**Comment by jerrinot** [comment]
what about including something like this:
```
//TODO: This has been added to preserve binary compatibility in Hazelcast 3.x. See <link> for explanation. This should be removed in Hazelcast 4.x release. 
```

**Comment by jerrinot** [comment]
a test would be nice. 

It should be quite easy to write a testing class using ASM. See this: https://github.com/jerrinot/experiments-asm/blob/master/src/main/java/uk/co/rockstable/experiments/asm/FooBarFactory.java  - it defines a class with 2 argument-less methods with the same name, but with a different return type.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java (hunk 1 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java]: agree with @jerrinot

Comment (ajermakovics) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java]: indeed, I like that. means it'll be easier to navigate the call hierarchy in IDE as well

Comment (ajermakovics) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java]: On the other hand we set things accessible in many other places so can't see it making a difference

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java [hunk 2]
```diff
@@ -169,6 +170,16 @@ public static Comparable extractValue(Object object, String attributeName) throw
         return (Comparable) createGetter(object, attributeName).getValue(object);
     }
 
+    public static <T> T invokeMethod(Object object, String methodName) throws RuntimeException {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: Looks like this `ReflectionHelper ` is targeted to use with queries. Maybe you can create a global utility class for methods like `invokeMethod`.

Comment (ajermakovics) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: It is, but the methods themselves are quite generic

Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: did not get your point.

**Comment by Danny-Hazelcast** [comment]
any news on this PR ?


https://hazelcast-l337.ci.cloudbees.com/view/Stabilizer/job/Hazelcast-nightly-stabilizer/29/console
is failing 

```
cause=java.lang.AbstractMethodError: com.hazelcast.stabilizer.tests.map.helpers.MapStoreWithCounter.loadAllKeys()Ljava/lang/Iterable;
	at com.hazelcast.map.impl.MapStoreWrapper.loadAllKeys(MapStoreWrapper.java:102)
	at com.hazelcast.map.impl.mapstore.BasicMapStoreContext.loadInitialKeys(BasicMapStoreContext.java:222)
	at com.hazelcast.map.impl.mapstore.BasicMapStoreContext.access$200(BasicMapStoreContext.java:53)
	at com.hazelcast.map.impl.mapstore.BasicMapStoreContext$TriggerInitialKeyLoad.call(BasicMapStoreContext.java:285)
	at com.hazelcast.map.impl.mapstore.BasicMapStoreContext$TriggerInitialKeyLoad.call(BasicMapStoreContext.java:281)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)

```

**Pull request merged**